### ISSUE TITLE
Are command buffers reusable?

### ISSUE THREAD

**Issue by benvanik** [gpuweb/gpuweb#1971]
The only mention in the spec I see about whether the same command buffer can be submitted multiple times is a `TODO: reusability flag?` here: https://www.w3.org/TR/webgpu/#command-encoder-creation

I was hoping they are and the flag would just be a hint to implementations that they are one-shot, but `GPUCommandBuffer.executionTime` makes me question this: that seems to assume a command buffer will only be submitted once as otherwise it's racy/undefined if the same command buffer is submitted multiple times in the same submit, etc. If there's a plan to add reusability/command buffer nesting/etc I'd probably flag that as being a risk as it'll make it really difficult to extend the behavior of the command buffer so long as it's there.

I noticed that there are GPURenderBundles which may take the place of reusable command buffers (as there are no secondary/nested command buffers, which is where reuse is most effective), however there's no matching GPUComputeBundle that could be used when performing compute. Has a GPUComputeBundle been discussed or is the feeling that it's not worth amortizing the API cost of recording the bundles for compute given that GPUComputePassEncoder doesn't have transfer operations and heavy compute usage often has a mix of transfer and compute operations.

**Comment by kainino0x** [comment]
Command buffers are not reusable, which is a restriction in (all?) underlying APIs - and iirc we've also been told that Vulkan reusable command buffers don't work out as well in practice as one might think.

As you mention, we have GPURenderBundle, which has been also designed around the constraints of underlying apis like [ID3D12GraphicsCommandList](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nn-d3d12-id3d12graphicscommandlist). We've certainly considered something like GPUComputeBundle, although at the moment I'm personally interested in the possibility of a fully-general reusable "secondary" command buffer - that would probably _not_ map to native API concepts and instead just result in re-recording in the backends.

**Comment by benvanik** [comment]
Makes sense! We've mostly found the same thing from an execution-time perspective and just treat them as a way to memoize recording. It's nice being able to record a few dozen commands to execute cached command buffers + update some parameter memory instead of re-recording a thousand commands per frame (and running all the logic to compute them) - especially if having to hop wasm->v8->etc boundaries for each API call like in the browser :)

**Comment by kvark** [comment]
> Command buffers are not reusable, which is a restriction in (all?) underlying APIs - and iirc we've also been told that Vulkan reusable command buffers don't work out as well in practice as one might think.

It's technically only a restriction in Metal. Vulkan and D3D12 are fine with re-usability (but Vulkan can benefit from non-reusable command buffers).

The reason why we only have bundles for rendering is because render passes are what can contain thousands or tens of thousands draw calls. You don't see this in compute work, and WebGPU currently over-synchronizes dispatches anyway, so making a lot of them is not scaleable.

**Comment by benvanik** [comment]
Nit: _you_ don't see this in compute work today - but _we_ do - and we'd love for more of these kind of workloads to work in WebGPU so others could play as well :)

A data point here is the recently added CUDA graph support - which are basically just command buffers (cached DAG of commands that can be submitted multiple times). We haven't yet done the deep work on whether there are execution performance benefits (current driver versions doing more sophisticated optimization work on the graphs vs imperative stream recording) but they certain drop our overhead (5-10ms -> 100us) in submitting work. I suspect they have some good reasoning for adding such a large feature to their API given that it's not a win on hand-authoring/usability.

Once we have more (any) of our workload running on WebGPU we can profile the time spent in resource tracking, barrier insertion, etc in the implementations and see how that would change if it could be reused. Will be happy to post some benchmarks (and ask for help from implementations in how best to profile them :).

**Comment by kvark** [comment]
Thanks for the info! I'm excited to see your case informing our future work on allowing unsynchronized dispatches within a compute pass.
